oauthenticator/google.py

```diff
@@ -338,71 +338,54 @@ def _service_client_credentials(self, scopes, user_email_domain):
 
         return credentials
 
-    def _service_client(self, service_name, service_version, credentials, http=None):
+    async def _setup_credentials(self, user_email_domain):
         """
-        Return a configured service client for the API.
+        Set up the oauth credentials for Google API.
         """
+        credentials = await self._service_client_credentials(
+            scopes=[f"{self.google_api_url}/auth/admin.directory.group.readonly"],
+            user_email_domain=user_email_domain,
+        )
+
         try:
-            from googleapiclient.discovery import build
+            from google.auth.transport import requests
         except:
             raise ImportError(
-                "Could not import googleapiclient.discovery's build,"
+                "Could not import google.auth.transport's requests,"
                 "you may need to run 'pip install oauthenticator[googlegroups]' or not declare google groups"
             )
 
-        self.log.debug(
-            f"service_name is {service_name}, service_version is {service_version}"
-        )
-
-        return build(
-            serviceName=service_name,
-            version=service_version,
-            credentials=credentials,
-            cache_discovery=False,
-            http=http,
-        )
-
-    def _setup_service(self, user_email_domain, http=None):
-        """
-        Set up the service client for Google API.
-        """
-        credentials = self._service_client_credentials(
-            scopes=[f"{self.google_api_url}/auth/admin.directory.group.readonly"],
-            user_email_domain=user_email_domain,
-        )
-        service = self._service_client(
-            service_name='admin',
-            service_version='directory_v1',
-            credentials=credentials,
-            http=http,
-        )
-        return service
+        request = requests.Request()
+        credentials.refresh(request)
```

Comment:
I'm not sure what goes on here, but i think this may involve API communication etc, making python blocked. Not great, but we can live with it i figure.

Reply:
This PR is an incremental improvement no matter what.

Im not sure what goes on here though, is it making API requests etc?

Reply:
It is indeed making an API request, and I've been trying to find a way to make it async will no luck. Here's are my initial thoughts re what we should do after reading through more docs and learning more about how the authenticator works:

1. Go back to having the creds be set in the object instead of setting them up and refreshing the token for each login
2. Add methods that check whether the creds exist and the token is valid/non-expired and refresh them using the synchronous call if not. See https://google-auth.readthedocs.io/en/latest/reference/google.auth.credentials.html#google.auth.credentials.Credentials.

This will be the flow:

1. When the first user logs in, creds are setup and a token is retrieved (synchronously)
2. When user 1 + N user logs in, the authenticator checks that the token is still valid. If so, no need to make another _synchronous_ call to get it, use it so fetch groups _asynchronously_. If not, send the _synchronous_ request to refresh the token.

This way we reduce the number of synchronous calls to nearly 0 while also accounting for [potential token expiration](https://developers.google.com/identity/protocols/oauth2#expiration).

Reply:
Latest commit implements the above, @consideRatio.

Also, the test with multiple repeated logins several users went well, seems like the general async fetching of groups works well.

Reply:
Its very elegant to do it like this, nice! I wonder if a tweak may be needed though, as discussed in https://github.com/jupyterhub/oauthenticator/pull/764/commits/389d4b2a723fe48ddce215f8c20210124d0fc363#r1788133777.

I don't get what credentials we have, but since we reference `user_email_domain`, it makes me think there may be an issue when we have mutliple users with different domain in their email. I really don't know though - but from a shallow code inspection and little knowledge on the involved kinds of credentials I suspect it could be an issue.

Reply:
Ha, @consideRatio this totally flew past me. All our hubs are single domain, so it didn't occur to me. Latest commit stores creds in a dictionary of `{domain: credentials}`. These creds, or in this case the token, can be used multiple times for authenticating users of the particular domain.